### DISCUSSION TITLE
RAG RestAPI Calls

### DISCUSSION THREAD

**Discussion by crawfordg** [open-webui/open-webui#3991] [Q&A]
Using postman, is there an API call that will generate a chat using a document collection?

**Comment by pnmartinez** [comment]
Really interested into this too, with no success so far.

This question is reproduced by many, such as: 
 https://github.com/open-webui/open-webui/discussions/3804
https://github.com/open-webui/open-webui/discussions/3465

**Comment by pnmartinez** [comment]
Solved!! ✅

Obtained the call's structure **from the `Network` tab from the Browser DevTools.**. The key is to add a `files` field to the payload, as this is what triggers the RAG flow; see next line of the codebase https://github.com/open-webui/open-webui/blob/8d257ed596b7172613144e4ed5ed9c0cedc65b7f/backend/main.py#L583 

To pass your file's data, look at the call on the **Network tab on the DevTools when sending a RAG message**  on the chat on Open WebUI. Then update the following Python script with your data, or get it properly through other API calls.

```py
import requests

url = "http://localhost:3000/ollama/api/chat"

headers = {
    "Accept": "application/json",
    "Accept-Language": "en-EN,en;q=0.9",
    "Authorization": "Bearer"+ YOUR_TOKEN,
    "Connection": "keep-alive",
    "Content-Type": "application/json",
    "Cookie": "token="+YOUR_TOKEN,
    "Origin": "http://localhost:3000",
    "Referer": "http://localhost:3000/",
    "Sec-Fetch-Dest": "empty",
    "Sec-Fetch-Mode": "cors",
    "Sec-Fetch-Site": "same-origin",
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/127.0.0.0 Safari/537.36",
    "sec-ch-ua": "\"Not)A;Brand\";v=\"99\", \"Google Chrome\";v=\"127\", \"Chromium\";v=\"127\"",
    "sec-ch-ua-mobile": "?0",
    "sec-ch-ua-platform": "\"Windows\""
}

# Payload
data = {
    "stream": False, # if True, returns response a sequence of partial JSONs
    "model": YOUR_MODEL,
    "messages": [
        {"role": "user", "content": YOUR_PROMPT}
    ],
    "options": {},
    "files": [
      # to test this with your files, better get your file's data by checking the Network tab on the DevTools when sending
      #  a RAG message thorugh Open WebUI
        {
            "type": "file",
            "collection_name": "9a45cfaf40249db8cd1..6d9e8b9e9b283f20d1eee2d0a4ec",
            "name": "NAME",
            "title": "FILENAME.pdf",
            "filename": "FILENAME.pdf",
            "content": {},
            "user_id": "61f2fa5f-7...d-ed8b0bb0af4c",
            "timestamp": 1723377796,
            "status": "processed"
        }
    ],
    "session_id": "l-_J1...kcZ-a9fZAAAD",
    "chat_id": "aeaacae5-74..-bf81-039fec1281f4",
    "id": "d0a3b65d-e....9-8892-6091100a0674"
}

response = requests.post(url, headers=headers, json=data)

print(response.status_code)
print(response.text)
```

**Reply by energy29uk** [reply]
This was a great fix by pnmartinez and working great until I updated to the latest version of openwebui 0.3.30,  now the API responses no longer returns any citations, only the message itself. Would anyone know if this is a new bug or simply a change in how the API works ? if I revert to my old version 0.3.16 the citations in the JSON response appear again.

**Reply by VicentePina7210** [reply]
I am curious if you ever found a fix to this.
